File: src/07.greedy/02.subArrayGreedy/13.402.go
package sub_array_greedy

func RemoveKdigits(num string, k int) string {
	stack := []byte{}
	for _, v := range num {
		ch := byte(v)
		for k > 0 && len(stack) > 0 && stack[len(stack)-1] > ch {
			stack = stack[:len(stack)-1]
			k--
		}

		if v == '0' && len(stack) == 0 {
			continue
		}
		stack = append(stack, ch)
	}

	for k > 0 && len(stack) > 0 {
		stack = stack[:len(stack)-1]
		k--
	}

	if len(stack) == 0 {
		return "0"
	}

	return string(stack)
}

File: src/07.greedy/02.subArrayGreedy/13.402.ts
export function removeKdigits(num: string, k: number): string {
    const stack = [];

    for (let i = 0; i < num.length; i++) {
        while (k > 0 && stack.length && stack[stack.length - 1] > num[i]) {
            stack.pop();
            k -= 1;
        }

        // 处理前导 0, 为 0 时, 并且 stack 为空时, 不入栈
        if (num[i] === "0" && stack.length === 0) {
            continue;
        }
        stack.push(num[i]);
    }

    // 如果 k 还存在,
    while (k > 0 && stack.length > 0) {
        stack.pop();
        k--;
    }

    if (stack.length === 0) {
        return "0";
    }

    return stack.join("");
}
